crypto/merklekeystore/keystore.go

```diff
@@ -79,6 +81,11 @@ type (
 	}
 )
 
+// MaxValidPeriod defines the longest validity period allowed for key registration,
+// to prevent the merklekeystore tree depth being larger than 16
+var MaxValidPeriod = (1<<16)*config.Consensus[protocol.ConsensusFuture].CompactCertRounds - 1
```

Comment:
you want to calculate this dynamically using an init function.

Reply:
What is the advantage of initializing it in init()? Is it just a convention?

Reply:
The calculation you've used above is wrong - it assumes that the value in `protocol.ConsensusFuture` is the highest. This might not hold true in the future. In `config/consensus.go`, we have a `checkSetAllocBounds` which is called via an init function, which calculate the various max parameters by checking *all* the consensus params.

Reply:
@tsachiherman is this correct?

init() function is called when the package is initialized. Example of it is the one in config/consensus.go
As I understand, Tsachi is suggesting to have an init function, which loops through all the `ConsensusProtocols`, and picks the largest value of `CompactCertRounds`.
e.g.
```
func init() {
        compactCertRounds = 0
        for _, p := range config.Consensus {
                if p. CompactCertRounds > compactCertRounds
                       compactCertRounds = p. CompactCertRounds
                }
        }
        MaxValidPeriod = (1<<16)*compactCertRounds - 1
}
```

Reply:
Given that this is a consensus protocol max value, why don't we "port" this one into `consensus.go` ?
We already have all the infrastructure there.

Reply:
I'm not sure why we would need the max of all `CompactCertRounds` though, this is not the behavior I would expect when generating the StateProof keys;
The number of keys generated (and thus the depth of the merkle tree that will be built) is derived from `validityPeriod/CompactCertRounds`, so in the case where we lower `CompactCertRounds` in future consensus version the number of keys will increase accordingly - and will be bigger than the 2^16 limit we require. 
This limit defines the longest validity period that can be requested, so for a smaller `CompactCertRounds` I would require a smaller value.

Reply:
I could be wrong, but I believe that we use it during the decoding of the msgp. We want to create a backward-compatible limit. This number doesn't have to be "accurate", but it should provide a reasonable "protection" against potential malicious msgp-content.

Reply:
This variable is moved to be a local non-exported variable inside a function.